util/equality/equality_test.go

```diff
@@ -136,3 +141,106 @@ func TestJobConfigChanged(t *testing.T) {
 		}
 	}
 }
+
+func TestDeploymentConfigChanged(t *testing.T) {
+	zero := int32(0)
+	two := int32(2)
+
+	testCases := []struct {
+		description string
+		mutate      func(deployment *appsv1.Deployment)
+		expect      bool
+	}{
+		{
+			description: "if nothing changes",
+			mutate:      func(_ *appsv1.Deployment) {},
+			expect:      false,
+		},
+		{
+			description: "if replicas is changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Replicas = &zero
+			},
+			expect: true,
+		},
+		{
+			description: "if selector is changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Selector = &metav1.LabelSelector{}
+			},
+			expect: true,
+		},
+		{
+			description: "if the container image is changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Template.Spec.Containers[0].Image = "foo:latest"
+			},
+			expect: true,
+		},
+		{
+			description: "if a volume is changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Template.Spec.Volumes = []corev1.Volume{
+					{
+						Name: "foo",
+						VolumeSource: corev1.VolumeSource{
+							HostPath: &corev1.HostPathVolumeSource{
+								Path: "/foo",
+							},
+						},
+					}}
+			},
+			expect: true,
+		},
+		{
+			description: "if container commands are changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Template.Spec.Containers[0].Command = []string{"foo"}
+			},
+			expect: true,
+		},
+		{
+			description: "if container args are changed",
+			mutate: func(deploy *appsv1.Deployment) {
+				deploy.Spec.Template.Spec.Containers[0].Args = []string{"foo", "bar", "baz"}
+			},
+			expect: true,
+		},
```

Comment:
Can you add a test case like this?  https://github.com/openshift/cluster-ingress-operator/blob/c6b66d029f59e71b8d5e345c39beaa819a54d204/pkg/operator/controller/ingress/deployment_test.go#L846-L861

The reason I ask for that test case is kind of subtle, so I'll try to explain:  `DesiredDeployment` defines probes like this:
```go
		LivenessProbe: &corev1.Probe{
			Handler: corev1.Handler{
				HTTPGet: &corev1.HTTPGetAction{
					Path: "/healthz",
					Port: intstr.IntOrString{IntVal: int32(8000)},
				},
			},
		},
```
When the operator creates the resource, the API sets default values for the unspecified fields (`TimeoutSeconds`, `PeriodSeconds`, `SuccessThreshold`, and `FailureThreshold`).

Later the operator gets the deployment and checks the desired and actual deployments.  The API has set values for `TimeoutSeconds` etc., and these values aren't in the deployment that `DesiredDeployment` builds, so `DeploymentConfigChanged` returns true, which causes `updateDeploymentIfNeeded` to try to update to the desired deployment, every time the reconciliation loop runs.  At least this is what I suspect will happen in contour-operator because we have had similar things happen in other operators.  Possible solutions are (a) to define the comparison to treat omitted values the same as the corresponding default values or (b) to define explicit values for any field for which the API sets a default value.

Reply:
@Miciah that for catching this. I missed it in my testing, but I'll push a PR to fix the issue by setting the defaults in the desired `Deployment` resource.